TableauConnectedApp, TransifexOrganization, and OpenmrsImporter models: AES CBC encryption Migration

Jtang-1 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
N/A

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Continuation of the work initiated by https://github.com/dimagi/commcare-hq/pull/35403
[Jira Ticket](https://dimagi.atlassian.net/browse/USH-5223)

Introduces a solution for these code scan alerts:
https://github.com/dimagi/commcare-hq/security/code-scanning/295
https://github.com/dimagi/commcare-hq/security/code-scanning/296

This PR migrates `TableauConnectedApp`'s `secret_value`, `TransifexOrganization`'s `api_token`, and `OpenmrsImporter`'s `password.  For values that are AES ECB encrypted, this migration reencrypts it with AES CBC. For values that are not encrypted, this migration encrypts it with AES CBC. In both cases, the prefix "$aes-cbc$" is prepended to the encrypted value and saved. It leaves blank values as is.

On Prod, `TableauConnectedApp` has 8 items, `TransifexOrganization` has only 1 item, and `OpenmrsImporter` has 20 items.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
The TableauConnected model is only relevant to [embedded_tableau](https://www.commcarehq.org/hq/flags/edit/embedded_tableau/) and [tableau_user_syncing](https://www.commcarehq.org/hq/flags/edit/tableau_user_syncing/) FF

The `OpenmrsImporter` is only relevant to [openmrs_integration](https://www.commcarehq.org/hq/flags/edit/openmrs_integration/)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
locally tested and will test on staging.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
There exists tests that reencrypt_ecb_to_cbc_mode decrypts to the same plaintext value.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [X] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->
The migration in this PR is related to this [past PR ](https://github.com/dimagi/commcare-hq/pull/35665) and should not be merged until that PR is deployed.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: create migration to reencrypt secret value for TableauConnectedApp

Jtang-1 committed:
feat: migrates api_token for transifex to use cbc encryption

Jtang-1 committed:
feat: handle error with decrypting existing token value

Since the value can not be decrypted, it means it means it's broken in its
current state so just set it as blank.

Jtang-1 committed:
refactor: for reversion, filter only for value that have been migrated, indicated by the 'aes-cbc' prefix

Jtang-1 committed:
bug: fix migration dependency and add to migrations lock

Jtang-1 committed:
Merge branch 'master' into jt/cbc-migration-pt2

jingcheng16 commented on corehq/apps/geospatial/migrations/0009_geoconfig_api_key_cbc_encryption.py:
```diff
@@ -0,0 +1,52 @@
+from django.db import migrations
```
This is from a merged PR?

Jtang-1 replied:
Oops must have accidentally added this when i did `git add .` Thanks for catching this! deleted the file in this commit [b2dae2e](https://github.com/dimagi/commcare-hq/pull/35666/commits/b2dae2ed30fab286eecdb5a76afa3af6cd31fa0c)

jingcheng16 commented on migrations.lock:
```diff
@@ -606,6 +606,7 @@ geospatial
  0006_geoconfig_max_cases_per_user_and_more
  0007_geoconfig_max_case_distance_and_more
  0008_geoconfig_flag_assigned_cases
+ 0009_geoconfig_api_key_cbc_encryption
```
This is from a merged PR?

Jtang-1 replied:
Yes that was a mistake, deleted here [b2dae2e](https://github.com/dimagi/commcare-hq/pull/35666/commits/b2dae2ed30fab286eecdb5a76afa3af6cd31fa0c)

jingcheng16 reviewed (commented):
Looks like you accidentally introduced code in other branch into this branch. Otherwise it looks good.

Jtang-1 committed:
delete accidently added file. This file is part of another PR

MartinRiese commented on corehq/apps/translations/migrations/0010_transifex_organization_api_token_encryption.py:
```diff
@@ -0,0 +1,51 @@
+    )
+
+    for org in transifex_orgs_to_update:
+        org.api_token = reencrypt_cbc_to_ecb_mode(org.api_token,
```
Is there a chance this will throw an exception if `reencrypt_ecb_to_cbc_mode` on line 23 fails?

Jtang-1 replied:
By fails do you mean via the `AesEcbDecryptionError` exception or another unexpected failure?

In both cases, I wouldn't expect `reencrypt_cbc_to_ecb_mode` to throw an exception. This is because we are only updating values that are prefixed with `aes-cbc` (`ALGO_AES_CBC`) via the filter below. If the value has the prefixed, it means that it has already successfully been migrated which means its reversion will also be successful.

```
transifex_orgs_to_update = TransifexOrganization.objects.filter(
        api_token__startswith=f'${ALGO_AES_CBC}$'
    )
```

MartinRiese reviewed (changes requested):
Just want the question answered before this is getting merged. Otherwise I am fine with Jing's approval

Jtang-1 committed:
Merge branch 'jt/cbc-password-migration' into jt/cbc-migration-pt2

Jtang-1 committed:
feat: migrates openmrs to use cbc encrypted password

the original password is not prefixed but the migrated password will be prefixed with aes-cbc

Jtang-1 committed:
bug: prefix is already appended by reencrypt function

Jtang-1 commented:
closing and reopening so tests will run

jingcheng16 reviewed (approved):
Looks good! Just one thing regarding PR, force push should be avoided after a PR has been reviewed. It can change the commit history, then make the reviewer confused and don't know if the previous commits that they approved are still the same, although it seems like you didn't change it.

sravfeyn commented:
FYI, when I tried to deploy staging (which included this), there was a migration error

```
Traceback (most recent call last):
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/corehq/sql_db/management/commands/migrate_multi.py", line 66, in handle
    job.get()
  File "src/gevent/greenlet.py", line 805, in gevent._gevent_cgreenlet.Greenlet.get
  File "src/gevent/greenlet.py", line 373, in gevent._gevent_cgreenlet.Greenlet._raise_exception
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/gevent/_compat.py", line 49, in reraise
    raise value.with_traceback(tb)
  File "src/gevent/greenlet.py", line 908, in gevent._gevent_cgreenlet.Greenlet.run
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/corehq/sql_db/management/commands/migrate_multi.py", line 43, in migrate_db
    call_command('migrate', *args, **call_options)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/core/management/__init__.py", line 194, in call_command
    return command.execute(*args, **defaults)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/corehq/sql_db/management/commands/migrate.py", line 17, in handle
    result = super().handle(*args, **options)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/corehq/util/django_migrations.py", line 121, in _inner
    return migration_fn(*args, **kwargs)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/corehq/apps/reports/migrations/0021_tableauconnectedapp_use_aes_cbc_encryption.py", line 30, in migrate_tableau_connected_app_secret_value
    connected_app.save()
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/base.py", line 814, in save
    self.save_base(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/base.py", line 877, in save_base
    updated = self._save_table(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/base.py", line 990, in _save_table
    updated = self._do_update(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/base.py", line 1054, in _do_update
    return filtered._update(values) > 0
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/query.py", line 1231, in _update
    return query.get_compiler(self.db).execute_sql(CURSOR)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1984, in execute_sql
    cursor = super().execute_sql(result_type)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1562, in execute_sql
    cursor.execute(sql, params)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/sentry_sdk/utils.py", line 1713, in runner
    return sentry_patched_function(*args, **kwargs)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/sentry_sdk/integrations/django/__init__.py", line 650, in execute
    result = real_execute(self, sql, params)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/cchq/www/staging/releases/2025-02-02_14.19/python_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.DataError: value too long for type character varying(64)", "stdout_lines": ["", "The following databases will be migrated:", " * default", " * auditcare", " * p1", " * p2", " * p3", " * p4", " * p5", " * proxy", " * repeaters", " * synclogs", "", "", "The following databases will be skipped:", " * ucr", "", "Started at 2025-02-02 14:38:54", "Processing [ ] 0/1 0% -:--:-- remaining", "Processing [.] 1/1 100% 0:00:00.000140 elapsed", "Processing [.] 1/1 100% 0:00:00.000170 elapsed", "Finished at 2025-02-02 14:38:54", "Elapsed time: 0:00:00", "", "======================= Error During Migration ======================="
```

Having looked at the migration in this PR, it might be that the encrypted-secret value is larger than 64 chars. 

I made a revert branch and added it to staging so that this won't block staging deploys. You should resolve this, or else it might block the production deploy too.
